### PR TITLE
Add /pm command to message_wnd

### PR DESCRIPTION
**Pull request by TheSilentOne1**
Add /pm command that sends a private message to a single player, as an alternative to selecting a player from the player list wnd.
Based on [#2503](https://github.com/freeorion/freeorion/pull/2503)

### CODE REVIEW

**Comment by o01eg** [comment]
1) How it will work if the player is disconnected?
2) Does server show it only to one player? If message will be processed by `ChatHistoryProvider.put_history_entity`, it will be shown public after server restart.

**Comment by TheSilentOne1** [comment]
1. In our last games, I sent some messages which don't seem to have been transmitted. So my guess is, the current message implementation was designed for and in its current form is probably suited only for players playing simultaneously.
2. I don't know, but if you can read `ChatHistoryProvider.put_history_entity`, we can easily test if messages become public after server restart. I can send you a private message on our current game by selecting you on PlayerListWnd (but probably both of us should be online at the same time to be sure?).

**Comment by o01eg** [comment]
@TheSilentOne1 I fixed disclosure of private messages in #2511 . Could you fix conflicts and rebase?

**Comment by TheSilentOne1** [comment]
Rebased.

File: UI/ChatWnd.cpp [hunk 1]
```diff
@@ -622,31 +622,79 @@ namespace {
...
+    }
+    else if (boost::iequals(command, "help")) {
+        *m_display += UserString("MESSAGES_HELP_COMMAND") + "\n";
+        m_display_show_time = GG::GUI::GetGUI()->Ticks();
+    }
+    else if (boost::iequals(command, "pm")) {
+        int player_id = ExtractPlayerID(params);
+        std::string message = ExtractMessage(params);
+
+        if (player_id != -1) {
```
Comment (geoffthemedio) [inline UI/ChatWnd.cpp]: replace `-1` with `Networking::INVALID_PLAYER_ID`

**Comment by geoffthemedio** [comment]
This doesn't work in the multiplayer lobby. `ClientApp::Players` returns an empty map, even though I've got two human clients connected to the server.

**Comment by geoffthemedio** [comment]
Works in-game, but it would probably be good to indicate in the message whether it was a pm or public message... like
````
[04 Aug 12:32:32] Human_Player [to All]: message to all players
[04 Aug 12:32:43] Human_Player [to OtherPlayer]: message just to one player
````

**Comment by geoffthemedio** [comment]
@TheSilentOne1 Any plans to do more for this?

**Comment by TheSilentOne1** [comment]
Yes, in the foreseeable future.

**Comment by geoffthemedio** [comment]
@TheSilentOne1 poke

**Comment by TheSilentOne1** [comment]
Soon, patience please :)

**Comment by geoffthemedio** [comment]
Typing `/pm playername` (with no message after) crashes the client. As does just `/pm`.

**Comment by TheSilentOne1** [comment]
Ouch. I was busy lately, but I'll have time to look into this and overally improve it later this week.

File: UI/ChatWnd.cpp [hunk 2]
```diff
@@ -611,42 +613,104 @@ void MessageWnd::OpenForInput() {
...
+        std::string recipient_names = "";
 
         if (recipients.empty()) {
             net.SendMessage(PlayerChatMessage(text));
         } else {
-            recipients.insert(HumanClientApp::GetApp()->PlayerID());   // ensure recipient sees own sent message
-            for (int recipient_id : recipients)
-                net.SendMessage(PlayerChatMessage(text, recipient_id));
+            for (int recipient_id : recipients) {
+                net.SendMessage(PlayerChatMessage(UserString("MESSAGES_WHISPER") + " " + text, recipient_id));
```
Comment (geoffthemedio) [inline UI/ChatWnd.cpp]: (Whisper) probably shouldn't be baked into the actual message text. Rather, use a string format substutution on the recipient client when displaying

File: UI/ChatWnd.cpp [hunk 3]
```diff
@@ -611,42 +613,104 @@ void MessageWnd::OpenForInput() {
...
+
+                auto player_it = players.find(recipient_id);
+                if (player_it != players.end())
+                    recipient_names += player_it->second.name + "; ";
+            }
+            recipient_names = recipient_names.substr(0, recipient_names.size() - 2);
+            net.SendMessage(PlayerChatMessage("[" + UserString("TO") + ": " + recipient_names + + "] " + text, HumanClientApp::GetApp()->PlayerID()));
```
Comment (geoffthemedio) [inline UI/ChatWnd.cpp]: Same here... the message metadata shouldn't be baked in on the sender's side

File: UI/ChatWnd.cpp (hunk 2 above)
Comment (TheSilentOne1) [reply UI/ChatWnd.cpp]: Could you give an example where or how that is done?

Comment (geoffthemedio) [reply UI/ChatWnd.cpp]: eg. https://github.com/freeorion/freeorion/blob/master/UI/CombatReport/GraphicalSummary.cpp#L358 or https://github.com/freeorion/freeorion/blob/master/UI/ChatWnd.cpp#L80 or search for `boost::io::str` for lots of examples.

File: UI/ChatWnd.cpp (hunk 3 above)
Comment (TheSilentOne1) [reply UI/ChatWnd.cpp]: Should be fixed.

**Pull request merged**